Source/WarpXEvolve.cpp

```diff
@@ -252,6 +222,173 @@ WarpX::EvolveEM (int numsteps)
+    ApplyFilterandSumBoundaryRho(fine_lev, PatchType::fine, 0, 2);
+    NodalSyncRho(fine_lev, PatchType::fine, 0, 2);
+
+    EvolveB(fine_lev, PatchType::fine, 0.5*dt[fine_lev]);
+    EvolveF(fine_lev, PatchType::fine, 0.5*dt[fine_lev], DtType::FirstHalf);
+    FillBoundaryB(fine_lev, PatchType::fine);
+    FillBoundaryF(fine_lev, PatchType::fine);
+
+    EvolveE(fine_lev, PatchType::fine, dt[fine_lev]);
+    FillBoundaryE(fine_lev, PatchType::fine);
+
+    EvolveB(fine_lev, PatchType::fine, 0.5*dt[fine_lev]);
+    EvolveF(fine_lev, PatchType::fine, 0.5*dt[fine_lev], DtType::SecondHalf);
+
+    if (do_pml) {
+        DampPML(fine_lev, PatchType::fine);
+        FillBoundaryE(fine_lev, PatchType::fine);
+    }
+
+    FillBoundaryB(fine_lev, PatchType::fine);
+
+    // ii)
+    PushParticlesandDepose(coarse_lev, curtime);
+    StoreCurrent(coarse_lev);
+    AddCurrentFromFineLevelandSumBoundary(coarse_lev);
+    AddRhoFromFineLevelandSumBoundary(coarse_lev, 0, 1);
+
+    EvolveB(fine_lev, PatchType::coarse, dt[fine_lev]);
+    EvolveF(fine_lev, PatchType::coarse, dt[fine_lev], DtType::FirstHalf);
+    FillBoundaryB(fine_lev, PatchType::coarse);
+    FillBoundaryF(fine_lev, PatchType::coarse);
+
+    EvolveE(fine_lev, PatchType::coarse, dt[fine_lev]);
+    FillBoundaryE(fine_lev, PatchType::coarse);
+
+    EvolveB(coarse_lev, PatchType::fine, 0.5*dt[coarse_lev]);
+    EvolveF(coarse_lev, PatchType::fine, 0.5*dt[coarse_lev], DtType::FirstHalf);
+    FillBoundaryB(coarse_lev, PatchType::fine);
+    FillBoundaryF(coarse_lev, PatchType::fine);
+
+    EvolveE(coarse_lev, PatchType::fine, 0.5*dt[coarse_lev]);
+    FillBoundaryE(coarse_lev, PatchType::fine);
+
+    // iii)
+    UpdateAuxilaryData();
+
+    // iv)
+    PushParticlesandDepose(fine_lev, curtime+dt[fine_lev]);
+    RestrictCurrentFromFineToCoarsePatch(fine_lev);
+    RestrictRhoFromFineToCoarsePatch(fine_lev);
+    ApplyFilterandSumBoundaryJ(fine_lev, PatchType::fine);
+    NodalSyncJ(fine_lev, PatchType::fine);
+    ApplyFilterandSumBoundaryRho(fine_lev, PatchType::fine, 0, 2);
+    NodalSyncRho(fine_lev, PatchType::fine, 0, 2);
+
+    EvolveB(fine_lev, PatchType::fine, 0.5*dt[fine_lev]);
+    EvolveF(fine_lev, PatchType::fine, 0.5*dt[fine_lev], DtType::FirstHalf);
+    FillBoundaryB(fine_lev, PatchType::fine);
+    FillBoundaryF(fine_lev, PatchType::fine);
+
+    EvolveE(fine_lev, PatchType::fine, dt[fine_lev]);
+    FillBoundaryE(fine_lev, PatchType::fine);
+
+    EvolveB(fine_lev, PatchType::fine, 0.5*dt[fine_lev]);
+    EvolveF(fine_lev, PatchType::fine, 0.5*dt[fine_lev], DtType::SecondHalf);
+
+    if (do_pml) {
+        DampPML(fine_lev, PatchType::fine);
+        FillBoundaryE(fine_lev, PatchType::fine);
+    }
+
+    FillBoundaryB(fine_lev, PatchType::fine);
+    FillBoundaryF(fine_lev, PatchType::fine);
+
+    // v)
+    RestoreCurrent(coarse_lev);
+    AddCurrentFromFineLevelandSumBoundary(coarse_lev);
+    AddRhoFromFineLevelandSumBoundary(coarse_lev, 1, 1);
```

Comment:
To my understanding, the second argument of function `AddRhoFromFineLevelandSumBoundary` is the component of the MultiFab on which the filter is applied. In part `ii)` of this loop, it is applied on component `0` only, and here it is applied on component `1`. Are these two `Rho` to push PMLs?

Reply:
For div e cleaning (i.e., EvolveF).